### PR TITLE
First branch

### PR DESCRIPTION
**Pull request by santosh-sivakumar**
I've added 2 files - one for the vector class, and one for the KMeans algorithm. Please look over

### CODE REVIEW

File: Vector.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		
+	Vector (List<Double> numbers) {
+			
+		numbers = numbers;
+			
+	}
+		
+	public double distance(Vector vector2) { // returns distance between two vectors
+		
+		double squared_distance = 0;
```
Comment (aravinho) [inline Vector.java]: need error checks here. What if the other vector is not of the same length as this one?  If that's the case, throw an Exception

File: Vector.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	
+	List<Double> numbers;
+		
+	Vector (List<Double> numbers) {
+			
+		numbers = numbers;
+			
+	}
+		
+	public double distance(Vector vector2) { // returns distance between two vectors
```
Comment (aravinho) [inline Vector.java]: nit: usually in methods like this that compare another object to "this" object, the name of the argument is `other`, rather than `vector2`.

File: Vector.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+			origin_coordinates.add(0.0);
+		}
+			
+		Vector origin = new Vector(origin_coordinates);
+			
+		return distance(origin);
+			
+	}
+}
+
```
Comment (aravinho) [inline Vector.java:45]: Looks like proper Java style conventions say to end a file with 2 blank lines

File: Vector.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+		for (int index = 0; index < numbers.size(); index++) {
+			double difference = numbers.get(index) - vector2.numbers.get(index);
+			squared_distance += (difference * difference);
+		}
+			
+		return Math.sqrt(squared_distance);	
+					
+	}
+	
+	public double magnitude() { // computes vector magnitude
```
Comment (aravinho) [inline Vector.java]: Your implementation is correct, but its needlessly expensive.  Yes, the magnitude of a vector is the distance from the origin, and it's nice that you were able to use the helper function, but in this implementation, you are looping once to build up a list, creating a new instance of the Vector class, and will loop again in the call to `distance`.  This function should just be a single loop that accumulates a running sum of squares.

Comment (aravinho) [inline Vector.java]: needs proper docstring

File: Vector.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import java.io.*;
+import java.util.*;
+import java.lang.Math; 
+
+public class Vector {
+	
+	List<Double> numbers;
+		
+	Vector (List<Double> numbers) {
+
```
Comment (aravinho) [inline Vector.java]: You have a lot of empty lines.  Follow this style guide: https://google.github.io/styleguide/javaguide.html

File: KMeans.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
+import java.io.*;
```
Comment (aravinho) [inline KMeans.java:1]: Don't use wildcard imports (c.f.https://google.github.io/styleguide/javaguide.html)

File: KMeans.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.*;
+import java.lang.Math; 
+
+public class KMeans {
```
Comment (aravinho) [inline KMeans.java:5]: Every class and function (at minimum, every `public` class and function) needs a proper docstring, with a brief description, as well as `@param` and `@return` tags.  See https://www.oracle.com/technetwork/articles/java/index-137868.html

File: KMeans.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+	private Vector midpoint (List<Vector> listofvectors) { //computes + returns a midpoint Vector from a given list of Vectors
```
Comment (aravinho) [inline KMeans.java:7]: Google Style guide recommends 100 character limits per line.  I actually can't read past the word "given" on my laptop.

Comment (aravinho) [inline KMeans.java:7]: parameters should be in `lowerCamelCase`. So, `listOfVectors`.  But actually, I'd recommend just `vectors` as the name of this param

File: KMeans.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+		
+		List<Double> mp_coordinates = new ArrayList<Double>();
+		double number_of_vectors = listofvectors.size();
```
Comment (aravinho) [inline KMeans.java:10]: I like underscores, but the Java community (and the Google style guide) uses `lowerCamelCase` for variable names. Use that, and I'd suggest `midpointCoordinates`

Comment (aravinho) [inline KMeans.java:10]: `number_of_vectors` --> `numVectors`.  And this should be an `int` not a `double`.  Also, you need error checks to make sure that all the vectors in the given list are of the same length.

File: KMeans.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+		double length_of_vector = listofvectors.get(0).numbers.size();
```
Comment (aravinho) [inline KMeans.java:11]: use `lowerCamelCase` for local variables (here and everywhere else).  And this should be `vectorLength`.

File: KMeans.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+public class KMeans {
+
+	private Vector midpoint (List<Vector> listofvectors) { //computes + returns a midpoint Vector from a given list of Vectors
+		
+		List<Double> mp_coordinates = new ArrayList<Double>();
+		double number_of_vectors = listofvectors.size();
+		double length_of_vector = listofvectors.get(0).numbers.size();	
+			
+		for (int dimension = 0; dimension < length_of_vector; dimension ++) {
```
Comment (aravinho) [inline KMeans.java:13]: nit, no space before `++`

File: KMeans.java (hunk 10 above)
Comment (aravinho) [inline KMeans.java:11]: what if someone passed a list with 0 vectors. You'd get a array out of bounds error from the `get(0)` call.  So at the top of this function (along with your checks that all vectors of the same length), you should add a check that at least one vector is actually provided

File: KMeans.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+				
+			coordinate /= number_of_vectors;
+			mp_coordinates.add(coordinate);
+				
+		}
+		return new Vector(mp_coordinates);
+		
+	}
+	
+	private int closestCenter (Vector start_vector, List<Vector> listofcenters) { // given a starting vector and list of cluster centers, returns closest center's index
```
Comment (aravinho) [inline KMeans.java:30]: character limit.  And these inline comments need to be replaced with proper Javadoc strings

File: KMeans.java [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+		
+		int closest_index = 0;
+		double distance_to_closest = start_vector.distance(listofcenters.get(0));
```
Comment (aravinho) [inline KMeans.java:33]: need error checks. make sure you weren't passed an empty list. Make sure all vectors are of the same length.

File: Vector.java (hunk 2 above)
Comment (aravinho) [inline Vector.java]: I think this should be a static method, rather than an instance method.  The signature is cleaner if it's
```
public static double distance(Vector vector1, Vector vector2) ...
```

File: KMeans.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+		
+		
+		for (int curr = 1; curr < listofcenters.size(); curr++) {
```
Comment (aravinho) [inline KMeans.java:36]: cleaner if you initialize `indexOfClosest = -1` and `distToClosest = -1.0`, and then start the loop from index 0.

File: KMeans.java [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+			if (distance < distance_to_closest) {
+				closest_index = curr;
+				distance_to_closest = distance;
+			}
+		}
+		
+		return closest_index;
+	}
+	
+	private boolean is_clustering_finished (List<Vector> old_centers, List<Vector> current_centers) { // computes normalized difference between cluster centers of 2 subsequent iterations [for all clusters], returns whether clustering is finished
```
Comment (aravinho) [inline KMeans.java:48]: I don't understand the logic here.  Consider the following case:
```
current_centers[0] = [100000000000000000, 1000000000000000000]
old_centers[1] = [100000000000000000, 100000000000000100]
```
distance_between_centers is 100 (which is pretty large), but `magnitude will be so large` that it will force `normalized_distance` under the threshold, and terminate the algorithm.

Comment (aravinho) [reply KMeans.java:48]: In general, why are you normalizing anything?  A better termination condition would use absolute comparsions, like "at least X% of the cluster centers are a distance < D away from their previous values", for some constants X and D.

File: KMeans.java [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+	
+		for (int index = 0; index < current_centers.size(); index++) {
+			double distance_between_centers = old_centers.get(index).distance(current_centers.get(index));
+			double magnitude = current_centers.get(index).magnitude();
+			double normalized_distance = distance_between_centers / magnitude;
+
+			
+			if (normalized_distance > 0.00000001) {
```
Comment (aravinho) [inline KMeans.java:56]: Constants like this should be specified in `CONSTANT_CASE` near the top of the file. (https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names)

File: KMeans.java [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	private List<List<Vector>> empty_clusters (int quantity) { // initializes List of desired number of clusters (each cluster is a list of vectors)
+	
+		List<List<Vector>> cluster_list = new ArrayList<List<Vector>>();
```
Comment (aravinho) [inline KMeans.java:65]: Why is one an `ArrayList` and one a `List`? Any particular reason?

File: KMeans.java [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+		
+		for (int count = 0; count < quantity; count++) {
+			List<Vector> cluster = new ArrayList<Vector>();
+			cluster_list.add(cluster);
+		}
+		
+		return cluster_list;
+	}
+			
+	public List<List<Vector>> cluster (List<Vector> unclustered_vectors, int number_of_clusters) { // full cluster algorithm
```
Comment (aravinho) [inline KMeans.java:75]: `number_of_clusters` --> `numClusters`

File: KMeans.java [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+	
+		List<List<Vector>> returnlist = new ArrayList<List<Vector>>(); // initializes cluster list to be returned
+		int number_of_vectors = unclustered_vectors.size();
```
Comment (aravinho) [inline KMeans.java:78]: error checks.  What if I pass an empty list.  What if I pass `quantity = -1`.  Simple errors like that should be checked for and caught early, and should crash loudly and with a clear error message.

File: KMeans.java [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+		
+		List<Vector> current_centers = new ArrayList<Vector>();
+		for (int count = 0; count < number_of_clusters; count++) { // chooses random centers to begin, adds to list of centers
```
Comment (aravinho) [inline KMeans.java:81]: Didn't you discover a problem with this approach?  Shouldn't the inital vectors just have random coordinates in the space `((min_x, max_x), (min_y, max_y), (min_z, max_z), ...)`?

File: KMeans.java [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+			int random_index = (int)(Math.random() * number_of_vectors);
+			current_centers.add(new Vector(unclustered_vectors.get(random_index).numbers));
+		}
+
+		List<Vector> old_centers = new ArrayList<Vector>(); // initializes old list of centers for isClusteringFinished method
+
+		do { // main loop, iterates through all vectors, finds closest center, adds vector to cluster with index corresponding to center's index in center list
```
Comment (aravinho) [inline KMeans.java:88]: nobody uses `do ... while ...` any more.  Use a standard while loop

File: KMeans.java [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+				}
+			}
+			
+		} while (! is_clustering_finished(old_centers, current_centers));
+
+		
+	return returnlist;
+	}
+}
+
```
Comment (aravinho) [inline KMeans.java:126]: 2 blank lines at EOF per style conventions

**Review by aravinho — CHANGES_REQUESTED** [review]
Logic is clean, and mostly easy to follow (except for your termination condition).
Main things to address:
* docstrings docstrings docstrings.  Every class and function should clearly describe its parameters and/or return values, and what the function does.  Use Javadoc formatting (https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names)
* style/conventions: you interchange between camel case, underscores, etc, all over the place.  I commented in some places, but make the fixes throughout. Stick to the Google Style Guide: https://google.github.io/styleguide/javaguide.html
* error checks.  Sanity check all params, catch simple errors early, fail loudly and with clear error messages.
* You never check in code without tests.  Write some unit tests for both classes.  Unit tests should check all corner cases (including tests that assert the proper error is thrown), and some simple normal cases.

btw, I reviewed the `Vector` class before the `KMeans` class so the comments may appear out of order.

File: KMeans.java (hunk 15 above)
Comment (santosh-sivakumar) [reply KMeans.java:48]: Good point, but what would be an appropriate D to capture all potential values?
